sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -461,6 +421,66 @@ object CollationTypeCoercion {
         else right
     }
   }
+
+  /**
+   * Throws an analysis exception if the new data type has indeterminate collation,
+   * and the expression is not allowed to have inputs with indeterminate collations.
+   */
+  private def checkIndeterminateCollation(expression: Expression, newDataType: DataType): Unit = {
+    if (shouldFailWithIndeterminateCollation(expression, newDataType)) {
+      expression.failAnalysis(
+        errorClass = "INDETERMINATE_COLLATION_IN_EXPRESSION",
+        messageParameters = Map("expr" -> toSQLExpr(expression)))
+    }
+  }
+
+  /**
+   * Returns whether the given expression which isn't allowed to have inputs with indeterminate
+   * collations has indeterminate collation.
+   */
+  private def shouldFailWithIndeterminateCollation(expression: Expression): Boolean = {
+    def getDataTypeSafe(e: Expression): DataType = try {
+      e.dataType
+    } catch {
+      case _: Throwable => NullType
+    }
+
+    expression.children.exists(child => expression.resolved &&
+      shouldFailWithIndeterminateCollation(expression, getDataTypeSafe(child)))
+  }
+
+  /**
+   * Returns whether the given expression should fail with indeterminate collation if it is cast
+   * to the given data type.
+   */
+  private def shouldFailWithIndeterminateCollation(
+      expression: Expression,
+      dataType: DataType): Boolean = {
+    !canContainIndeterminateCollation(expression) && hasIndeterminateCollation(dataType)
+  }
+
+  /**
+   * Returns whether the given data type has indeterminate collation.
+   */
+  private def hasIndeterminateCollation(dataType: DataType): Boolean = {
+    dataType.existsRecursively {
+      case IndeterminateStringType | StringTypeWithContext(_, Indeterminate) => true
+      case _ => false
+    }
+  }
+
+  /**
+   * Returns whether the given expression can contain indeterminate collation.
+   */
+  private def canContainIndeterminateCollation(expr: Expression): Boolean = expr match {
+    // This is not an exhaustive list, and it's fine to miss some expressions. The only difference
```

Comment:
Perhaps a comment with a guideline for engineers adding further expressions in the future would be helpful: In case the new expression can contain indeterminate collation, it should be added to the list here, to the best of knowledge. Still, even if that is not the case, there is still runtime handling that will ensure that the expression will fail accordingly (though the first path is preferable as it saves burning some extra cycles).

Reply:
The comment kind of explains that already, so I am not really sure how you propose to modify it?

Reply:
Well, I wouldn't use the wording that 'it's fine to miss some expressions', as I assume that we want to encourage engineer adding a new expression that cannot contain indeterminate collation to add it to the list below. An additional explanation can be provided that otherwise, if the expression is not added to the list but cannot contain indeterminate collation, it will still fail at runtime, as already pointed out.